src/typeEvaluator/typeEvaluate.ts

```diff
@@ -905,6 +1026,70 @@ function handleEverythingNode(_: EverythingNode, scope: Scope): TypeNode {
   } satisfies ArrayTypeNode<UnionTypeNode<ObjectTypeNode>>
 }
 
+function handleAndNode(node: AndNode, scope: Scope): TypeNode {
+  const left = walk({node: node.left, scope})
+  const right = walk({node: node.right, scope})
+  return mapConcrete(left, scope, (lhs) =>
+    mapConcrete(right, scope, (rhs) => {
+      const leftValue = booleanValue(lhs)
+      const rightValue = booleanValue(rhs)
+      if (leftValue.canBeTrue && rightValue.canBeTrue) {
+        if (leftValue.canBeFalse || rightValue.canBeFalse) {
+          if (leftValue.canBeNull || rightValue.canBeNull) {
+            return nullUnion({type: 'boolean'})
+          }
+
+          return {type: 'boolean'}
+        }
+
+        return {type: 'boolean', value: true}
+      }
+
+      // this will never resolve, but we check if the expressions returns false, null or an union.
+      if (leftValue.canBeFalse || rightValue.canBeFalse) {
+        if (leftValue.canBeNull || rightValue.canBeNull) {
+          return nullUnion({type: 'boolean', value: false})
+        }
+        return {type: 'boolean', value: false}
+      }
+
+      return {type: 'null'}
+    }),
+  )
+}
+
+function handleOrNode(node: OrNode, scope: Scope): TypeNode {
+  const left = walk({node: node.left, scope})
+  const right = walk({node: node.right, scope})
+  return mapConcrete(left, scope, (lhs) =>
+    mapConcrete(right, scope, (rhs) => {
+      const leftValue = booleanValue(lhs)
+      const rightValue = booleanValue(rhs)
+      if (leftValue.canBeTrue || rightValue.canBeTrue) {
+        if (leftValue.canBeFalse || rightValue.canBeFalse) {
+          if (leftValue.canBeNull || rightValue.canBeNull) {
+            return nullUnion({type: 'boolean'})
+          }
+
+          return {type: 'boolean'}
+        }
+
+        return {type: 'boolean', value: true}
+      }
```

Comment:
Isn't this wrong if both values are `[canBeTrue, canBeNull]` and `[canBeTrue, canBeNull]`? 🤔 

In which case the actual type should be `nullIUnion({type: 'boolean'})`, but here it would be detected as `{type: 'boolean', value: true}`?

I think potentially this can be easier expressed by defining the logic on BooleanInterpretation:

```js
function booleanOr(left: BooleanInterpretation, right: BooleanInterpretation): BooleanInterpretation {
  // One side is true => return `true`
  if (left.canBeTrue && !left.canBeFalse && !left.canBeNull) return left
  if (right.canBeTrue && !right.canBeFalse && !right.canBeNull) return right

  // One side is false => return the other side
  if (!left.canBeTrue && left.canBeFalse && !left.canBeNull) return right
  if (!right.canBeTrue && right.canBeFalse && !right.canBeNull) return left

  // Guaranteed to be boolean at least:
  if (!left.canBeNull && !right.canBeNull) return {canBeTrue: true, canBeFalse: true, canBeNull: false}

  // Fallback to the most permissive:
  return {canBeTrue: true, canBeFalse: true, canBeNull: true}
}
```

I think this covers the main cases.

And then `and` becomes quite symmetric:

```js
function booleanAnd(left: BooleanInterpretation, right: BooleanInterpretation): BooleanInterpretation {
  // One side is true => return that side
  if (left.canBeTrue && !left.canBeFalse && !left.canBeNull) return right
  if (right.canBeTrue && !right.canBeFalse && !right.canBeNull) return left

  // One side is false => return `false`
  if (!left.canBeTrue && left.canBeFalse && !left.canBeNull) return left
  if (!right.canBeTrue && right.canBeFalse && !right.canBeNull) return right

  // Guaranteed to be boolean at least:
  if (!left.canBeNull && !right.canBeNull) return {canBeTrue: true, canBeFalse: true, canBeNull: false}

  // Fallback to the most permissive:
  return {canBeTrue: true, canBeFalse: true, canBeNull: true}
}
```